proposals/0046-astro-env.md

```diff
@@ -0,0 +1,367 @@
+
+# Detailed Design
+
+## Terminology
+
+- **Public variable**: variable replaced by its value at build time. Uses `import.meta.env` under the hood
+- **Secret variable**: variable retrieved at runtime, never part of the bundle. Uses runtime specific features, like `process.env` or `Deno.env.get()`
+- **Server variable**: variable available server-side only
+- **Client variable**: variable available client-side only
+
+## Astro config
+
+A new `env.schema` property is added to `AstroUserConfig`. Its type looks like so:
+
+```ts
+type StringField = {
+  type: "string"
+  optional?: boolean
+  default?: string
+}
+
+type NumberField = {
+  type: "number"
+  optional?: boolean
+  default?: number
+}
+
+type BooleanField = {
+  type: "boolean"
+  optional?: boolean
+  default?: boolean
+}
+
+type EnvFieldType = StringField | NumberField | BooleanField
+
+type PublicClientEnvFieldMetadata = {
+  scope: "client"
+  access: "public"
+}
+
+type PublicServerEnvFieldMetadata = {
+  scope: "server"
+  access: "public"
+}
+
+type SecretServerEnvFieldMetadata = {
+  scope: "server"
+  access: "secret"
+}
+
+type EnvSchema =
+  | Record<`PUBLIC_${string}`, PublicClientEnvFieldMetadata & EnvFieldType>
+  | Record<string, (PublicServerEnvFieldMetadata | SecretServerEnvFieldMetadata) & EnvFieldType>
+
+type AstroUserConfig = {
+  env?: {
+    schema: EnvSchema
+  }
+}
+```
+
+Since there are various combinations possible that can make choosing confusing, we provide a `envField` helper that uses chaining, mirroring other APIs like Astro DB or zod:
+
+```ts
+import { envField } from "astro/config"
+
+// { scope: "client", access: "public", type: "number", default: 4321 }
+envField.public().client().number({ default: 4321 })
+
+// { scope: "server", access: "secret", type: "string" }
+envField.secret().server().string()
+```
+
+Note that a variable is required by default, and can be made optional with `optional: true`
+
+## Integrations
+
+The above way of declaring the schema allows integrations to declare their own constraints from inside `astro:config:setup` without any special handling:
+
+```ts
+import { envField } from "astro/config"
+import type { AstroIntegration } from "astro"
+
+const integration = {
+  name: "supabase-integration",
+  hooks: {
+    "astro:config:setup": (params) => {
+      params.updateConfig({
+        env: {
+          schema: {
+            PUBLIC_SUPABASE_URL: envField.public().client().string()
+          }
+        }
+      })
+    }
+  }
+} satisfies AstroIntegration
+```
+
+## Public variables
+
+Public variables are checked at some point between `astro:config:setup` and `astro:server:setup`. Since we don't/won't support many data structures, custom validators are used for the validation. They will also be used for secret variables at runtime.
+
+Some names must be reserved, like `SSR` to avoid conflicts with `import.meta.env.SSR`.
```

Comment:
I don't think we should reserve any names. Someone might very well have that as an environment variable.

If it's due to `import.meta.env`, I'm not sure we should be using that in the implementation. IIRC, `import.meta.env` falls back to runtime on the server. And if we're saying that you must explicitly call `private()` to get a runtime variable, then we shouldn't use import.meta.env.